### ISSUE TITLE
Q1: qualify BF16 numerics and validate last-exited KV independently

### ISSUE THREAD

**Issue by hsliuustc0106** [ThinkFlowLab/vllm-rlt#4]
## Precision requirements — 2026-09-12

For new work, **BF16 is the primary GPU inference, kernel-validation, quality, and performance target**. Use BF16 weights, ordinary activations, and KV storage; specify accumulation independently and retain FP32 in numerically sensitive reductions/intermediates. Full-model FP32 is an optional diagnostic/reference configuration, not the default prerequisite. See the [precision policy](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/precision-policy.md).

BF16 profiling, matched optimization A/B, and quality data collection may proceed alongside the original Q1 reference diagnosis after each run's functional, finiteness, and ownership prerequisites pass. Report observations, reference fidelity, and qualification separately. Preserve exact cache/data-movement/RNG invariants and metadata-only A/B behavior; differing arithmetic paths use independently justified, predeclared BF16 numerical and decision/quality criteria. Universal bitwise or top-1 equality is not a blanket mixed-precision requirement, and near-tie exemptions cannot be invented after a run.

For dtype choices, follow explicit original Ouro paper guidance first, then the pinned release configuration and inference code. The [source audit](https://github.com/hsliuustc0106/vllm-lt/blob/c9e646ace2d8198a52e522ea4d71129f1b143c58/docs/paper-notes.md#original-ouro-paper-and-precision-guidance) distinguishes published FP32 operations from backend/project choices; the paper has no per-kernel inference recipe.

This amendment supersedes conflicting dtype/eligibility requirements below for successor work. All dated contracts, raw results, original failures, and historical acceptance decisions remain unchanged. New criteria need a separately frozen contract; no benchmark or qualification is claimed by this edit.

The first-pass two-dtype case table and exact-top-1 gates below remain the historical diagnostic specification. A successor contract targets BF16, documents per-operation accumulation/casts, keeps the pinned official and independent LAST-EXITED checks, and uses FP32 only for a stated diagnostic question. Distinguish same-state local error from accumulated-state drift and report the old failures under their original policy. An eager reference rounding difference alone does not justify promoting every kernel to FP32.

Parent roadmap: #2. **Q1** can proceed in parallel with the M1 benchmark/profiler issue #3. Proposed interfaces below describe future implementation, based on preview commit `fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7`.

## Problem and goal

Real-checkpoint FP32 validation passed on the recorded inputs, but BF16 failed its predeclared `atol=0.25, rtol=0.02` logit check. At depth four the recorded maximum errors were 0.5625 for Torch versus dense, 0.625 for Triton versus dense, and 0.28125 for Triton versus Torch. Matching short generated continuations does not resolve that failure.

Establish an independent incremental reference for **last-exited KV**, locate the numerical divergence across layers and loops, and make an evidence-backed BF16 qualification decision. Distinguish an implementation/cache defect from arithmetic differences and decision sensitivity. A documented failure or inconclusive diagnosis must remain visible; it is not BF16 qualification.

## Goals

- Validate the model equations, per-depth KV history, cumulative gate, and scheduler behavior using independent reference execution.
- Compare BF16 implementations on identical fixed token histories before studying live continuations; add FP32 for a separately stated arithmetic-sensitivity question.
- Produce bounded layer/loop diagnostics and a reproducible classification of the original BF16 failure.
- Record a precise numerical acceptance decision and any focused corrective work required. M1 collects BF16 evidence alongside this diagnosis; full-model FP32 is optional.

## Non-goals

- Performance tuning, benchmark speed claims, graphs, asynchronous routing, new allocation policies, or another model.
- GSM8K/task-quality evaluation or BF16 deployment/default promotion solely from numerical smoke results; quality belongs to [Q2 #8](https://github.com/hsliuustc0106/vllm-lt/issues/8).
- Silently widening the original tolerance, changing the oracle to match the implementation, or treating scheduler self-consistency as independent validation.
- A generic model-validation framework or a second production engine. The reference remains small, serial, and diagnostic.

## Constraints and experiment envelope

Use the pinned real model/tokenizer revision `574fa66cb8bf5abdc979642d01cf2b79b16bfab1`; verify local config, weight and tokenizer content manifests. Freeze source SHA/modification hashes, exact device IDs, environment, arithmetic flags, fixture order, reference revision, and comparison policy. Preserve the existing [failure record](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/docs/validation.md).

Begin with tiny CPU state-machine/model tests without CUDA discovery. The real-model suite starts with **16 deterministic text fixtures**, four at each prompt length 16/64/128/256, with **eight frozen subsequent input tokens** per fixture. Record exact IDs, tokenizer/template provenance, and hashes. Avoid a full cross-product of every diagnostic option.

Eight continuation inputs produce **nine prediction points**: one after prefill and eight after `advance`. Set native `max_tokens=9`. Record actual logits/argmax before replacing the first eight sampled IDs with the frozen next inputs; the ninth prediction has no subsequent forward pass. Forced-depth fixtures contain nine output-depth entries, with index zero equal to four. Agreement of overridden input IDs is not top-1 prediction agreement.

The historical initial plan contains two dtype passes (FP32 and BF16). Each fixture uses one independent serial oracle, native Torch serial, native Triton serial, and Triton refill/no-refill packed execution on the same histories. Fix packed groups and chunk/page sizes before execution. Cover fixed depth plus an explicit forced-depth pattern containing shallow-to-deeper transitions; a separately bounded four-fixture live-gate check exercises actual threshold decisions at 0.7/minimum two loops. Fixed-depth behavior also receives a bounded comparison against reviewed, pinned official model code; its adaptive output selection is not a last-exited-history oracle.

Before device work, write an exact case table rather than multiplying those descriptions implicitly. Define hypothesis, changed variable for each comparison, controls, numeric criteria, memory estimate, and stop conditions. Allow one disjoint feasibility set (at most one fixture per dtype), followed by **one deterministic pass per declared case**, explicitly replacing the performance two-run default. Limit detailed tensor dumps to the first two preselected diagnostic fixtures plus at most two first failures selected by deterministic case/boundary order; capture them within the planned pass. Start with a two-hour overall reservation cap and ten minutes per case; the outer cap wins. Do not add reruns or cases silently.

Run every device operation through the verified scheduler with exact IDs and a descriptive note/timeout. Reuse the prepared environment and loaded weights where possible, preflight simultaneous reference/engine memory, and clean up only task-owned requests/processes/reservations. Stop device work on invalid history, nonfinite values, device failure, or cleanup failure; preserve partial evidence. A source fix or new numerical hypothesis requires a separately recorded comparison plan.

### 1. Independent reference ownership

Add a functional reference under the existing model-reference area, with validation harness code separate from runtime execution. It may read immutable config/weights and use PyTorch tensor primitives. It must not call native `recurrent()`, native norm/RoPE/attention helpers, `KVCacheManager`, `Scheduler`, or `LLMEngine`.

Proposed private validation interfaces:

```python
class SerialOuroOracle:
    def __init__(self, config: OuroConfig, weights: Mapping[str, Tensor], *, capacity: int): ...
    def prefill(self, prompt_ids: Sequence[int], *, observer: DiagnosticSink | None = None) -> TokenTrace: ...
    def advance(self, input_token_id: int, *, policy: ExitPolicy,
                forced_depth: int | None = None,
                observer: DiagnosticSink | None = None) -> TokenTrace: ...
    def snapshot_kv(self, *, positions: Sequence[int]) -> KVSnapshot: ...
    def close(self) -> None: ...
```

One oracle instance owns one request. Its dense cache has explicit `[depth, physical_layer, position, kv_head, head_dim]` K/V storage plus initialized-slot metadata, with a fixed capacity from the case plan. No paged addressing or allocator is shared with the engine. Diagnostic histories must not accidentally alias reusable tensor buffers.

`prefill` executes every prompt token at all four loops and returns the final prompt token's trace/logits for output index zero. `advance` consumes one supplied token at the next position and returns the trace that predicts the following output; it does not choose the next input. The validation harness owns teacher forcing or sampling and must label which mode is used. `close` releases oracle-owned state without clearing shared caches.

### 2. Last-exited invariants

- At each layer/depth, write the current token's K/V before attention; read only initialized causal positions including the current token.
- Keep the token position and RoPE phase fixed across loops. Apply end-of-loop normalization before both the next loop and the gate.
- Accumulate real conditional hazards at every executed loop, including those before the minimum exit depth. Threshold 1.0 explicitly selects full depth.
- After exit at one-based depth `e`, copy that token's final **per-layer K/V** into its slots at depths `e+1 ... 4`. Preserve shallower computed histories and neighboring tokens; never copy final hidden state as KV or recompute skipped depths.
- Output zero records depth four. Output index `g >= 1` consumes generated token `g-1` at position `prompt_length + g - 1`. The last sampled output needs no further KV write.
- `forced_depth` fixes executed history for arithmetic/cache comparisons while retaining real gate computation. Free-running gate decisions are checked separately; forced routing is not evidence of gate agreement.

Tiny tests must exercise adjacent tokens exiting at 2 then 4, mixed physical-layer states, block/chunk boundaries, missing initialized prefixes, fragmented engine pages, cancellation/reuse, and isolation across requests. Compare engine populated KV values with the dense oracle through a validation-only snapshot adapter; retain each component's own addressing rules.

### 3. Case, policy, and diagnostic records

Use versioned JSON (`schema_version: 1`, `artifact_type`) and stable case IDs. Reject malformed lengths, unsupported versions, duplicate IDs, invalid token/depth values, and nonfinite configuration numbers before execution.

| Interface | Required contract |
| --- | --- |
| `ValidationCase` | Case ID, prompt IDs, frozen continuation input IDs, history mode (`teacher_forced`/`live_gate`), forced depths when applicable, dtype/backend/schedule/chunk settings, expected capacity, reference provenance, and fixture hashes. |
| `ComparisonPolicy` | Policy ID/hash, per-dtype and per-boundary absolute/relative bounds, finiteness requirements, top-1 and gate/exit rules, near-tie reporting rules, and rationale/provenance. Every acceptance bound is fixed before its comparison. |
| `TokenTrace` | Input position, upcoming output index, executed depths, per-loop gate logits/probabilities/CDF, selected depth, final logits, and optional selected boundary summaries. Depth conventions are explicit: serialized exit depths are one-based. |
| `DiagnosticSink` | Receives immutable boundary metadata and bounded summaries; full tensors are copied only for planned snapshots. It never controls execution or changes tensor dtype/arithmetic. |
| `ComparisonRecord` | Case/comparison IDs, boundary `(position, depth, layer, operation)`, max and RMS error, reference RMS, finite counts, top-1 IDs/top-two margins where relevant, gate/CDF difference and threshold distance, and pass/fail reason under the named policy. |
| `QualificationReport` | Manifest/input hashes, completed/missing cases, original failure reproduction, earliest observed divergence, competing explanations, isolated evidence, decision (`qualified_on_declared_suite`, `failed`, `inconclusive`), limitations, and links to focused corrective work. |

Inspect attention inputs/outputs, per-layer residual outputs, loop-normalized hidden states, gate logits and LM logits. Measure candidate and reference on identical histories and dtype for implementation fidelity; label BF16-versus-FP32 sensitivity comparisons separately. Preserve raw reference scale and top-two margins so a small error near a decision boundary is visible.

Retain the original logit bounds: FP32 `atol=0.001, rtol=0.0001`; BF16 `atol=0.25, rtol=0.02`. Do not infer bounds for hidden/KV/gate tensors from logit bounds. Any additional bounds need independent justification in the frozen policy. If a bound cannot yet be justified, mark that diagnostic non-qualifying rather than silently treating it as a pass. A revised policy creates a new identified comparison and retains results under the old policy.

### 4. Harness and artifacts

Extend the existing validation command or add a small companion command; choose one coherent CLI rather than duplicating checkpoint loading. Proposed operations are `probe`, `run`, and `report`, with a case-plan JSON path, prepared checkpoint path, and new output directory. Probe/report are CPU-only; run is reservation-scoped. The plan resolves all cases, policy hashes, dump limits, reference code dependencies, and file-content verification before execution.

Artifacts include the resolved plan, content/environment manifest, per-case JSON records, bounded tensor snapshots, failures/cleanup records, and a Markdown qualification report. Reference loading/compatibility, downloads, and preparation are separate from inference. Review and pin official custom model code/dependencies before using that adapter; never substitute an unpinned upstream implementation if compatibility fails.

Stream summaries and release per-case activation/cache state to bound memory. The official adapter is used only for equivalent fixed-depth checks. A live-gate divergence must record the first differing decision and stop claiming subsequent logits share identical histories; use teacher forcing for a controlled arithmetic follow-up.

Elementwise candidate-reference errors require matching tensor values, not two independently computed summaries. Predeclare a separate byte limit for transient matching-boundary buffers, bounded spooling, or lockstep comparison; account for request/depth ordering before pairing values and release them after comparison. This working-memory budget is distinct from the small number of persisted diagnostic dumps. Report missing comparisons if the bound is exceeded.

## Acceptance criteria / 验收标准

- [ ] **AC-Q1-01 — Freeze an executable contract.** Before device execution, publish the resolved case table and content-hashed manifest: all 16 fixtures (four per declared prompt length), two dtypes, exact reference/backend/schedule/history assignments, packed groups, nine prediction points per teacher-forced case, and the bounded official/live-gate checks. Every comparison has named required versus diagnostic-only gates, independently justified bounds, and expected record counts. Preserve the stated feasibility, one-pass, tensor-buffer/dump, per-case, and two-hour budgets. Validation rejects unresolved or contradictory plans.

- [ ] **AC-Q1-02 — Establish reference independence and invariants.** Provide a source-review checklist demonstrating that the oracle shares only immutable weights/configuration and tensor primitives, with no native compute/cache/scheduler calls. Passing tiny CPU tests must independently check layer-specific skipped-depth KV propagation, shallow-to-deeper adjacent tokens, causal initialization, chunk/page boundaries, fragmentation, request isolation, cancellation/reuse, cumulative hazards below minimum depth, and threshold-one behavior. Existing engine/reference tests and lint/format checks pass; CPU checks perform no CUDA discovery.

- [ ] **AC-Q1-03 — Produce comparable raw evidence.** Every required case has complete, uniquely keyed results for its declared boundaries, using identical supplied token histories and correct request/position/depth correspondence. Record actual logits/argmax before teacher-forcing substitutions; nine prediction points consume only eight continuation inputs. Include elementwise max/RMS errors, scales, finiteness, populated KV comparisons, gate/CDF distances, and actual decision outcomes where required. Record the first live-history divergence and label later histories incomparable. Forced IDs/depths are never counted as independent prediction/gate agreement.

- [ ] **AC-Q1-04 — Preserve the original numerical question.** Retain results against FP32 `atol=0.001, rtol=0.0001` and BF16 `atol=0.25, rtol=0.02`. Any revised policy has its own predeclared identity, rationale, and report; it does not replace the old failure. Provide reproduction evidence or an explicit investigation of non-reproduction. Hidden/KV/gate bounds must be separately justified. Unspecified required bounds, missing comparisons, or near-tie exceptions invented after observing results cannot yield qualification.

- [ ] **AC-Q1-05 — Reach a supported diagnosis.** The reviewed report identifies the earliest observed failing boundary, separates same-dtype fidelity from BF16/FP32 sensitivity, and links isolated evidence for its classification. `qualified_on_declared_suite` requires every required numerical and behavioral gate to pass. A conclusive `failed` decision requires a documented counterexample to a required contract and sufficient controlled evidence to localize the failure being diagnosed; merely describing a difference as “rounding” is insufficient. Unlocalized divergence, unsupported explanation, or incomplete required coverage remains `inconclusive`.

- [ ] **AC-Q1-06 — Verify completeness and cleanup.** An offline report regenerates decisions from preserved records and reconciles planned/completed/failed/missing counts without dropping cases. Link source/input hashes, commands, reservation/device controls, bounded snapshots, raw comparisons, and cleanup records showing zero task-owned active requests/KV reservations after recoverable completion. Device/cleanup failures and budget exhaustion remain visible; no replacement trials occur without a new recorded plan.

**Closure rule:** close only when AC-Q1-01 through AC-Q1-06 pass and the reviewed diagnosis is qualified or conclusively failed. A completed negative diagnosis can close Q1 while leaving BF16 unqualified; an inconclusive diagnosis or missing evidence keeps it open. Record the decision and dependent blockers on roadmap #2. Neither closing Q1 nor passing its numerical suite promotes BF16 by default or establishes task quality; those claims still require Q2 #8.

References: [independent full-sequence equations](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/vllm_lt/models/reference.py), [gate/lifecycle order](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/vllm_lt/engine/llm_engine.py), and [reference comparability](https://github.com/hsliuustc0106/vllm-lt/blob/fa7a2ca5b833f2dd0f3e8db53ce111a7406496f7/docs/paper-notes.md).
